File: app/src/main/java/org/wen/gank/base/MvpFragment.java
package org.wen.gank.base;

import android.os.Bundle;
import android.support.annotation.CallSuper;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

/**
 * created by Jiahui.wen 2017-07-08
 */
public abstract class MvpFragment<V extends MvpView, P extends MvpPresenter<V>> extends Fragment
    implements MvpView {

    private ViewDelegate<V, P> viewDelegate;
    private boolean firstLoad = false;

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        viewDelegate = new ViewDelegate<V, P>() {
            @Override
            protected P createPresenter() {
                return MvpFragment.this.createPresenter();
            }
        };
        viewDelegate.onCreate(savedInstanceState);
    }

    protected abstract P createPresenter();

    protected abstract int getLayoutRes();

    @SuppressWarnings("unchecked")
    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container,
        @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(getLayoutRes(), container, false);
        initView(view);
        viewDelegate.attachView((V) this);
        return view;
    }

    protected abstract void initView(View view);

    @CallSuper
    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        checkLazyLoad();
    }

    @CallSuper
    @Override
    public void setUserVisibleHint(boolean isVisibleToUser) {
        super.setUserVisibleHint(isVisibleToUser);
        checkLazyLoad();
    }

    private void checkLazyLoad() {
        if (firstLoad) return;
        if (getUserVisibleHint() && getView() != null) {
            firstLoad = true;
            onLazyLoad();
        }
    }

    protected void onLazyLoad() {

    }

    @Override
    public void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        viewDelegate.onSaveInstanceState(outState);
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        viewDelegate.detachView();
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        viewDelegate.onDestroy();
    }

    protected P getPresenter() {
        return viewDelegate.getPresenter();
    }
}

File: app/src/main/java/org/wen/gank/mvp/MvpFragment.kt
package org.wen.gank.mvp

import android.os.Bundle
import android.support.annotation.CallSuper
import android.support.v4.app.Fragment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup

/**
 * created by Jiahui.wen 2017-07-08
 */
abstract class MvpFragment<V : MvpView, P : MvpPresenter<V>> : Fragment(), MvpView {

  private lateinit var viewDelegate: ViewDelegate<V, P>
  private var firstLoad = false

  protected abstract var layoutRes: Int

  protected val presenter: P
    get() = viewDelegate.presenter

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    viewDelegate = object : ViewDelegate<V, P>() {
      override fun createPresenter(): P {
        return this@MvpFragment.createPresenter()
      }
    }
    viewDelegate.onCreate(savedInstanceState)
  }

  protected abstract fun createPresenter(): P

  override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                            savedInstanceState: Bundle?): View? {
    return inflater.inflate(layoutRes, container, false)
  }

  protected abstract fun initView(view: View)

  @Suppress("UNCHECKED_CAST")
  @CallSuper
  override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
    super.onViewCreated(view, savedInstanceState)
    initView(view)
    viewDelegate.attachView(this as V)
    checkLazyLoad()
  }

  @CallSuper
  override fun setUserVisibleHint(isVisibleToUser: Boolean) {
    super.setUserVisibleHint(isVisibleToUser)
    checkLazyLoad()
  }

  private fun checkLazyLoad() {
    if (firstLoad) return
    if (userVisibleHint && view != null) {
      firstLoad = true
      onLazyLoad()
    }
  }

  protected open fun onLazyLoad() {

  }

  override fun onSaveInstanceState(outState: Bundle) {
    super.onSaveInstanceState(outState)
    viewDelegate.onSaveInstanceState(outState)
  }

  override fun onDestroyView() {
    super.onDestroyView()
    viewDelegate.detachView()
  }

  override fun onDestroy() {
    super.onDestroy()
    viewDelegate.onDestroy()
  }
}
